zniper.1
.\"
.\"  zniper.1 - zniper manual
.\"  
.\"  Copyright (c) 2003 Claes M. Nyberg <cmn@fuzzpoint.com>
.\"  All rights reserved, all wrongs reversed.
.\"      
.\"  Redistribution and use in source and binary forms, with or without
.\"  modification, are permitted provided that the following conditions
.\"  are met:
.\"
.\"  1. Redistributions of source code must retain the above copyright
.\"     notice, this list of conditions and the following disclaimer.
.\"  2. Redistributions in binary form must reproduce the above copyright
.\"     notice, this list of conditions and the following disclaimer in the
.\"     documentation and/or other materials provided with the distribution.
.\"  3. The name of author may not be used to endorse or promote products
.\"     derived from this software without specific prior written permission.
.\"      
.\"  THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
.\"  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
.\"  AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
.\"  THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
.\"  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
.\"  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
.\"  OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\"  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
.\"  OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF 
.\"  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"     
.\" $Id: zniper.1,v 1.2 2004-09-08 07:55:32 cmn Exp $
.\"

.TH ZNIPER 1 "Jan 2004" "zniper version 1.0" " "
.SH NAME
.B zniper
\- Display and reset TCP connections

.SH SYNOPSIS
.B zniper
.RB [ -hnpsvV ]
.RB [ -b 
.IB color ]
.RB [ -B    
.IB color ]
.RB [ -f      
.IB color ]
.RB [ -i
.IB iface ]
.RB [ -l
.IB file ]

.SH DESCRIPTION
.B zniper
displays TCP connections seen by the first available interface, or the one specified 
with 
.B -i
option. The connections displayed can be reseted by injecting RST packets into the
stream. Connections initiated before 
.B zniper
was started is also detected (unless
.B -s
option is used on the command line), these are displayed marked as 'Old' since the client
sending the initial SYN packet is not known.
 
.PP
.SH COMMANDLINE OPTIONS

.BI  "-b " color
.RS 4
Background color, defaults to black.
.RE

.BI  "-B " color
.RS 4
Border color, defaults to green.
.RE

.BI  "-f " color
.RS 4
Foreground color, defaults to white.
.RE

.B -h
.RS 4
Display short summary of commandline options.
.RE 

.BI  "-i " iface
.RS 4
Specify the interface to listen on.
.RE 

.BI  "-l " file
.RS 4
Log status information to file
.IR file .
A nice way of logging detected connections.
.RE 

.B -n
.RS 4
Do not attempt to resolve hostnames.
.RE

.B -s
.RS 4
Require the initial SYN packet for a connection to be displayed.
.RE

.B -V
.RS 4
Print version and exit.
.RE


.B -p
.RS 4
Do not put the interface in promiscous mode.
Note that the interface might be in promiscous mode
for some other reason, why you might receive "all" packets
seen anyway.
.RE 

.PP
.SH INTERACTIVE COMMANDS

.B c
.RS 4
Decrease size of connection window.
.RE

.B C
.RS 4
Increase size of connection window.
.RE

.B i
.RS 4
Toggle display of idle connections.
.RE

.B p
.RS 4
Sort by source port, smallest first.
.RE

.B P
.RS 4
Sort by source port, descending order.
.RE

.B r
.RS 4
Toggle resolving of hostnames.
.RE

.B k
.RS 4
Attempt to kill the highlighted connection by injecting a RST packet
in each direction. Note that holding down the key will result in an
attempt of killing the next highlighted connection.
.RE

.B v
.RS 4
Decrease verbose level.
.RE

.B V
.RS 4
Increase verbose level.
.RE

.B x
.RS 4
Toggle status window visibility
.RE

.B +
.RS 4
Highlight the connection above the currently highlighted.
.RE

.B -
.RS 4
Highlight the connection below the currently highlighted.
.RE


.PP
.SH AUTHOR
Claes M. Nyberg
.RS 0
.IR <cmn@fuzzpoint.com>
.RE


zniper.c
/*
 *  zniper.c  - zniper main file
 *
 *  Copyright (c) 2002 Claes M. Nyberg <cmn@fuzzpoint.com>
 *  All rights reserved, all wrongs reversed.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions
 *  are met:
 *
 *  1. Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *  2. Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in the
 *     documentation and/or other materials provided with the distribution.
 *  3. The name of author may not be used to endorse or promote products
 *     derived from this software without specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
 *  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
 *  AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
 *  THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 *  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 *  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 *  OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 *  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 *  OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 *  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */ 

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <pthread.h>
#include "zniper.h"

/* Global variables */
struct options opt;

/* Locks */
extern pthread_mutex_t writestat;
extern pthread_mutex_t conn_tree;

void
cleanup(void)
{
	endwin();
}

void
usage(char *pname)
{
	printf("\nZniper - <cmn@fuzzpoint.com>\n");
	printf("Usage: %s [Option(s)]\n", pname);
	printf("\n Options:\n");
	printf("  -b color   - Background color\n");
	printf("  -B color   - Border color\n");
	printf("  -f color   - Foreground color\n");
	printf("  -h         - This help\n");
	printf("  -i iface   - Network interface\n");
	printf("  -l file    - Log status information to file\n");
	printf("  -n         - Do not attempt to resolve hostnames\n");
	printf("  -p         - Do not put the interface in promiscuous mode\n");
	printf("  -s         - Require the initial SYN packet to display a connection\n");
	printf("  -v         - Verbose output, repeat to increase\n");
	printf("  -V         - Print \"%s\" and exit\n", VERSION);
	printf("\n");
}

int
get_color(char *str)
{
	if (!strcmp(str, "black"))
		return(COLOR_BLACK);
	else if (!strcmp(str, "red"))
		return(COLOR_RED);
	else if (!strcmp(str, "green"))
		return(COLOR_GREEN);
	else if (!strcmp(str, "yellow"))
		return(COLOR_YELLOW);
	else if (!strcmp(str, "blue"))
		return(COLOR_BLUE);
	else if (!strcmp(str, "magenta"))
		return(COLOR_MAGENTA);
	else if (!strcmp(str, "cyan"))
		return(COLOR_CYAN);
	else if (!strcmp(str, "white"))
		return(COLOR_WHITE);
	
	return(-1);
}

int
main(int argc, char *argv[])
{
	struct capture *cap;
	pthread_t sniff_thread;
	int flag;

	/* Default values */
	opt.bgc = COLOR_BLACK;
	opt.fgc = COLOR_WHITE;
	opt.boc = COLOR_GREEN;
	opt.conhp = 75;
	opt.promisc = 1;
	opt.sock_raw = -1;
	opt.resolve = 1;
	opt.grab = 1;
	opt.statw = 1;
	opt.idle = 1;
	opt.logfile = NULL;
	opt.usec = 1;

	while ( (flag = getopt(argc, argv, "f:b:B:l:i:nvVhps")) != -1) {
		switch(flag) {
				
			case 'b': opt.bgc = get_color(optarg); break;
			case 'B': opt.boc = get_color(optarg); break;
			case 'f': opt.fgc = get_color(optarg); break;
			case 'l':
				if (log_open(optarg) < 0)
					exit(EXIT_FAILURE);
				opt.logfile = optarg;
				break;
			
			case 's':
				opt.grab = 0;
				break;

			case 'h':
				usage(argv[0]);
				exit(EXIT_SUCCESS);
				break;
				
			case 'i':
				if (opt.iface != NULL)
					free(opt.iface);
				opt.iface = (u_char *)strdup(optarg);
				break;
			
			case 'n':
				opt.resolve = 0;
				break;
			
			case 'p':
				opt.promisc = 0;
				break;

			case 'v':
				opt.verbose++;
				break;

			case 'V':
				printf("%s\n", VERSION);
				exit(EXIT_SUCCESS);
				break;

			default:
				exit(EXIT_FAILURE);
		}
	}

	if (opt.bgc < 0 || opt.boc < 0 || opt.fgc < 0)
		opt.usec = 0;

	if (opt.verbose > 5)
		opt.verbose = 5;

	/* Open the raw socket to send packets from */
	if ( (opt.sock_raw = socket(PF_INET, SOCK_RAW, IPPROTO_RAW)) < 0) {
		perror("Error opening raw socket");
		exit(EXIT_FAILURE);
	}

	/* Open interface */
	if ( (cap = cap_open(opt.iface, opt.promisc)) == NULL)
		exit(EXIT_FAILURE);

	/* Filter out TCP packets */
	if (cap_setfilter(cap, "tcp") < 0)
		exit(EXIT_FAILURE);

	/* Revoke privs in case of setuid */
	seteuid(getuid());
	setuid(getuid());

	/* Create mutex objects */
	pthread_mutex_init(&writestat, NULL);
	pthread_mutex_init(&conn_tree, NULL);

	/* Set up screen */
    if (initscreen() < 0) {
		fprintf(stderr, "** Error: Failed to initialize screen\n");
		exit(EXIT_FAILURE);
	}
	drawscreen();

	writestatus(0, "Opened interface %s in %spromiscuous mode", 
		opt.iface, (opt.promisc == 1) ? "" : "non-");
	if (opt.logfile)
		writestatus(0, "Opened logfile %s", opt.logfile);
	writestatus(0, "Verbose level set to %d", opt.verbose);

	/* Create the sniff-thread */
	if (pthread_create(&sniff_thread, NULL, sniff, cap) != 0) {
		endwin();
		fprintf(stderr, "Error: Failed to create sniff thread\n");
		exit(EXIT_FAILURE);
	}

	atexit(cleanup);

	/* Enter interactive loop */
	iact();

	endwin();
	exit(EXIT_SUCCESS);
}
